examples/material-next-app-router-ts/src/components/Theme/ThemeRegistry/EmotionCache.tsx

```diff
@@ -0,0 +1,68 @@
+'use client';
+
+import * as React from 'react';
+import createCache from '@emotion/cache';
+import { useServerInsertedHTML } from 'next/navigation';
+import { CacheProvider as DefaultCacheProvider } from '@emotion/react';
+import type { EmotionCache, Options as OptionsOfCreateCache } from '@emotion/cache';
+
+export type NextAppDirEmotionCacheProviderProps = {
+  /** This is the options passed to createCache() from 'import createCache from "@emotion/cache"' */
+  options: Omit<OptionsOfCreateCache, 'insertionPoint'>;
+  /** By default <CacheProvider /> from 'import { CacheProvider } from "@emotion/react"' */
+  CacheProvider?: (props: {
+    value: EmotionCache;
+    children: React.ReactNode;
+  }) => React.JSX.Element | null;
+  children: React.ReactNode;
+};
+
+export function NextAppDirEmotionCacheProvider(props: NextAppDirEmotionCacheProviderProps) {
```

Comment:
```suggestion
// This implementation is taken from https://github.com/garronej/tss-react/blob/main/src/next/appDir.tsx
export function NextAppDirEmotionCacheProvider(props: NextAppDirEmotionCacheProviderProps) {
```

Ideally, this should come from emotion. @Andarist would the emotion time accept this as a contribution?

Reply:
Alright, we will use it for now in the example then, and I will keep an eye in the emotion issue. Thanks!

Reply:
Ah alright, that makes sense. @garronej would you like to make a contribution as you originally created it? :)

Reply:
Would be great if @garronej or @Andarist can share a summary of how this work with app router.

Reply:
Also why it is needed/what it accomplishes. I see some related exposition from NextJS but don't think it is terribly clear either for novice learners: https://nextjs.org/docs/app/building-your-application/styling/css-in-js#configuring-css-in-js-in-app

Reply:
If it's an issue you can make a PR on this @sgoodrow-zymergen

Reply:
> Would be great if @garronej or @Andarist can share a summary of how this work with app router.

Hello, the provider ensures that when the app is rendered on the server, the styles generated by Emotion get injected into the `<head>` element. 

Without the provider, Emotion defaults to its "zero config SSR mode", in which `<style />` tags are injected into the body of the document, right next to the element being styled. 

However, MUI is not compatible with Emotion's "zero config SSR mode". This has led to the recommended implementation with the Next Pages Router [here](https://github.com/mui/material-ui/blob/35d8bd6a0004667c370953c9ad7468197cca6a85/examples/material-next-ts/pages/_document.tsx) being quite complex. Ever since MUI started utilizing Emotion, it has required [the SSR advanced approach](https://emotion.sh/docs/ssr#advanced-approach) to be implemented, it's nothing new, it's the same old aprach implemented for the Next's App Router.  

A while back, I [proposed an RFC](https://github.com/mui/material-ui/issues/34575) suggesting an abstraction of this process. I've also initiated [a PR](https://github.com/emotion-js/emotion/issues/3054) on the Emotion repository to discuss transferring the TSS utilities. I'm absolutely prepared to contribute to this project. But as @Andarist pointed out, it would involve introducing a new experimental Emotion module, and he's better equipped to make the necessary decisions regarding that. 

@mnajdova @siriwatknp @sgoodrow-zymergen